src/basic/INDEXLIST-Cuda.cpp

```diff
@@ -0,0 +1,238 @@
+    }
+    __syncthreads();
+  }
+
+  if (last_block) {
+    for (int i = threadIdx.x; i < gridDim.x; ++i) {
+      block_readys[i] = 0u; // last block resets readys to 0 (for next kernel to reuse)
+    }
+  }
+
+  return count;
+}
+
+__device__ int get_block_id(unsigned* block_id_inc)
+{
+  __shared__ volatile unsigned s_block_id;
+  if (threadIdx.x == 0) {
+    s_block_id = atomicInc(block_id_inc, gridDim.x-1);
+  }
+  __syncthreads();
+  unsigned block_id = s_block_id;
+  __syncthreads();
+  return static_cast<int>(block_id);
+}
+
+__global__ void indexlist(Real_ptr x,
+                          Int_ptr list,
+                          Index_type* block_counts,
+                          Index_type* grid_counts,
+                          unsigned* block_readys,
+                          unsigned* block_id_inc,
+                          Index_type* len,
+                          Index_type iend)
+{
+  const int block_id = get_block_id(block_id_inc);
+
+  Index_type i = block_id * blockDim.x + threadIdx.x;
+  Index_type inc = 0;
+  if (i < iend) {
+    if (INDEXLIST_CONDITIONAL) {
+      inc = 1;
+    }
+  }
+
+  pair count = grid_scan(block_id, inc, block_counts, grid_counts, block_readys);
+
+  if (i < iend) {
+    if (count.first != count.second) {
+      list[count.first] = i;
+    }
+    if (i == iend-1) {
+      *len = count.second;
+    }
+  }
+}
+
+void INDEXLIST::runCudaVariant(VariantID vid)
+{
+  const Index_type run_reps = getRunReps();
+  const Index_type ibegin = 0;
+  const Index_type iend = getRunSize();
+
+  INDEXLIST_DATA_SETUP;
+
+  if ( vid == Base_CUDA ) {
+
+    INDEXLIST_DATA_SETUP_CUDA;
+
+    const size_t grid_size = RAJA_DIVIDE_CEILING_INT((iend-ibegin), block_size);
+
+    Index_type* len;
+    allocCudaPinnedData(len, 1);
+    Index_type* block_counts;
+    allocCudaDeviceData(block_counts, grid_size);
+    Index_type* grid_counts;
+    allocCudaDeviceData(grid_counts, grid_size);
+    unsigned* block_readys;
+    allocCudaDeviceData(block_readys, grid_size);
+    cudaErrchk( cudaMemset(block_readys, 0, sizeof(unsigned)*grid_size) );
+    unsigned* block_id_inc;
+    allocCudaDeviceData(block_id_inc, grid_size);
+    cudaErrchk( cudaMemset(block_id_inc, 0, sizeof(unsigned)) );
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      indexlist<<<grid_size, block_size, sizeof(Index_type)*block_size>>>(
+          x+ibegin, list+ibegin,
+          block_counts, grid_counts, block_readys, block_id_inc,
+          len, iend-ibegin );
+      cudaErrchk( cudaGetLastError() );
+
+      cudaErrchk( cudaDeviceSynchronize() );
+      m_len = *len;
+
+    }
+    stopTimer();
+
+    deallocCudaPinnedData(len);
+    deallocCudaDeviceData(block_counts);
+    deallocCudaDeviceData(grid_counts);
+    deallocCudaDeviceData(block_readys);
+    deallocCudaDeviceData(block_id_inc);
+
+    INDEXLIST_DATA_TEARDOWN_CUDA;
+
+  } else {
```

Comment:
No way to do this with a single kernel/loop with RAJA at the moment.

Reply:
It can't be done with RAJA Teams? What would we be missing?

Reply:
An abstraction for scans that can be done within a single RAJA kernel. We currently do with with 3 loops as can be seen in the INDEXLIST_3LOOP kernel in this PR.

Reply:
Ah gotcha, a device function of exclusive_scan_inplace, that users can call within their RAJA kernels? that would be useful!